lib/Parse/ParseGeneric.cpp

```diff
@@ -349,15 +355,37 @@ ParserStatus Parser::parseGenericWhereClause(
       }
 
       // Add the requirement
-      Requirements.push_back(RequirementRepr::getSameType(FirstType.get(),
-                                                      EqualLoc,
-                                                      SecondType.get()));
+      Requirements.push_back(RequirementRepr::getSameType(
+          FirstType, EqualLoc, SecondType.get(),
+          SourceRange(FirstTypeLoc, PreviousLoc)));
     } else {
       diagnose(Tok, diag::expected_requirement_delim);
       Status.setIsParseError();
       break;
     }
-    HasNextReq = consumeIf(tok::comma);
+    SourceLoc NextCommaLoc;
+    HasNextReq = consumeIf(tok::comma, NextCommaLoc);
+    SWIFT_DEFER { LastCommaLoc = NextCommaLoc; };
+
+    auto &req = Requirements.back();
+    auto removalRange = req.getRemovalRange();
+
+    if (!HasNextReq && LastCommaLoc.isInvalid()) {
+      // Include the where clause in the removal range for the first and only
+      // requirement.
+      removalRange.Start = WhereLoc;
+    }
+    if (!HasNextReq && LastCommaLoc.isValid()) {
+      // Include the previous comma in the removal range of the last
+      // requirement.
+      removalRange.Start = LastCommaLoc;
+    }
+    if (HasNextReq) {
+      // Include the next comma in the removal range of a middle requirement.
+      removalRange.End = NextCommaLoc;
```

Comment:
I think this means that one can have two requirements with overlapping removal ranges, e.g. `where X: Something, T: Any, U: Any`, both those last ones remove the trailing `,`. I don't know if the fixit-engine works correctly with this; might be worth including this sort of case in a test.

Reply:
That would mean that there's no fixits for the removal in `where T: Something, U: Any`, right? It does seem annoying we can't do multiple overlapping deletions, but it's probably better to miss the help in some cases instead of mangling the code.